### ISSUE TITLE
wrong path value returned by http_request::request_uri()

### ISSUE THREAD

**Issue by DenisZaharov** [microsoft/cpprestsdk#1410]
I have the next request:

`http://localhost:8888//api/v1/ws_auth`

I'm trying to handle this request but have got very strange values in return value of `http_request::request_uri`:

```
const web::uri& uri = request.request_uri() // request is the http_request parameter of callback
auto path = uri.path(); // "/v1/ws_auth"
auto host = uri.host(); // "api"
```
I guess the reason is double "//" in the uri after authority section, but a want to handle it correctly instead of having wrong host and path values. What's the right way to handle and fix that?

**Comment by garethsb-sony** [comment]
[RFC 3986 Section 3](https://tools.ietf.org/html/rfc3986#section-3) makes clear this is a tricky case because a path beginning `//` is valid within a URL which has a scheme and authority (host:port), but a URL that consists only of the path part **cannot** begin with `//`.

The client (browser or whatever) will have actually sent a request like
```
GET //api/v1/ws_auth HTTP/1.1
Host: localhost:8888
```
The `//api/v1/ws_auth` is known as the `request-target` in [RFC 7230](https://tools.ietf.org/html/rfc7230#section-5.3). The `request-target` in an HTTP 1.1 request is permitted to be an absolute URL or just the absolute path, so it's natural for a server (cpprestsdk...) to interpret it as a URL, but that breaks in your example.

**Comment by DenisZaharov** [comment]
Thanks for an answer.
But how can I handle this case on the server side? I suppose that clients can make any kind of requests, including incorrect requests like this. Will it be right to explicitly check the host value on emptiness because of the relative uri can't have host?

**Comment by garethsb-sony** [comment]
I think this may need a bug fix in the calls to `set_request_uri` inside the `http_listener` implementations. However, you may currently be able to work around it, by examining the components of the `request_uri` in your handler and reconstructing the full path from the `host`, `port` and `path`. Not ideal.

**Comment by garethsb-sony** [comment]
Or fix the client(s) not to make requests with the pesky extra `/` as I assume you already treat `http://localhost:8888/api/v1/ws_auth` identically?

**Comment by DenisZaharov** [comment]
Ok, thanks.
So, I will have to explicitly check the `request_uri` on the server side. I think I haven't to reconstruct the full path but just inform the client about an incorrect request.
I have also noticed that `request.relative_uri()` throws an exception when such kind of request is received. Maybe it's just a circumstance to make decision about incorrect request

**Comment by garethsb-sony** [comment]
Yes, if you can explain to clients that the correct URL is `http://localhost:8888/api/...` not `http://localhost:8888//api/...` then you just need to reject. If so, you may also want to reject a request_uri path that begins `//` also, in case the current behaviour is accepted as bug in C++ REST SDK code and fixed.
